File: Stacks/#4.Monotonic-Stacks/Python/NGE.py
"""
En este archivo veremos uno de los problemas as comunes que se pueden resolver de una forma optima utilizando pilas monotonas.
Este problema se trata  que dado un arreglo de elementos numericos, hay que hallar cual es el siguiente elemento mayor en el arreglo, de cada posicion. Si el elemento no tiene uno mayor a la derecha, sera -1 la respuesta de este. 
Veamos un ejemplo de como seria una entrada y salida de esto:

Input: arr[] = [4,5,2,25]
Output:  4 -> 5
		 5 -> 25
		 2 -> 25
		 25 -> -1

Ahora bien , lo que primero se pensaria al hacer este problema es iterar todo el arreglo por cada elemento de este, buscando proximo mayor de la posicion del arreglo en  la que nos encontremos. Sin embargo, en cuanto a eficiencia es un mal algoritmo ya que su complejidad temporal es cuadratica. 
Por ello, es mas conveniente usar otras estructuras como lo es la pila monotona o un mapa, los cuales aumentando la complejidad espacial a lineal, se puede reducir la complejidad temporal a lineal, ya sea con pilas monotonas o con un mapa.

Sin embargo, en este ocasion veremos como resolver este problema de la forma tipica, es decir, con dos ciclos, y con pilas monotonas, para mirar las diferencias entre ambos algoritmos, y asi ver una de las tantas aplicaciones que nos permite esta variante de las pilas. En este caso, implementaremos nuestra pila mediante una lista enlazada. 
"""

#Lo primero que debemos de hacer para crear un stack utilizando listas , es crear la estructura nodo
#Como se vio en la seccion de listas enlazadas, cada nodo contiene el dato a almcanear y el puntero al siguiente nodo
class Node:
    def __init__(self):  #Constructor vacio
        self.data  = 0 
        self.next = None
    
    def __init__(self,data):  #Constructor que recibe el dato que contendra el nodo
        self.data  = data
        self.next = None

#Ahora debemos de crear la estructura que llamabamos lista en aquella seccion, la cual contenia el nodo cabeza de la lista
#En este caso, llamaremos a esta estructura Stack, y en lugar de hacer un nodo encabezador , lo llamaremos un nodo top el cual siempre apuntara al ultimo nodo en entrar a la pila
class Stack:
    def __init__(self):
        self.top = Node(None)
        self.size = 0

    #Funcion para insertar en un Stack, es decir, la funcion push
    #Complejidad Temporal: O(1)
    def push(self, data):
        #Creamos el nuevo nodo a ser insertado
        new_node = Node(data)
        #Ahora, debemos de hacer que este nuevo nodo apunte al nodo Top del Stack, para ello, crearemos un nodo auxiliar que guarde la referencia inicial del nodo Top
        aux = self.top
        #Con la copia anterior creada, haremos que el nuevo nodo apunte a esta
        new_node.next = aux.next
        #Ahora, como el nodo top siempre debe de apuntar al ultimo nodo en entrar , haremos eso 
        self.top.next = new_node
        #Incrementamos el tamano del Stack
        self.size += 1
    
    #Funcion para retirar elemento de la pila. Por definicion de la pila, solo se puede retirar el elemento que esta en la cima de la pila, es decir, el ultimo entrar 
    #Complejidad Temporal: O(1)
    def pop(self):
        #Verificamos que el Stack no este vacio
        if self.size > 0 :
            #Recordemos que en las pilas, solo se puede retirar el elemento de la cima de esta, no se puede retirar en la posicion que uno desee
            #Para esto, sabemos que el nodo top siempre apunta a este nodo que esta en la cima que necesitamos retirar
            #Como este sera retirado, el nodo top debera de apuntar al nuevo nodo que ocupe la cima, es decir, apuntara al siguiente del nodo que esta en la cima actualmente
            #Primero lo que haremos sera guardar una referencia del nodo que se encuentra actualmente en la cima
            peek = self.top.next
            #Ahora, haremos que el nodo Top apunte a la nueva cima, es decir, el siguiente del nodo de la cima
            self.top.next = peek.next
            #Para desvincular totalmente el nodo que estaba en la cima, lo haremos apuntar a nulo
            peek.next = None
            #Ahora, se pueden realizar dos opciones, se puede retornar este nodo con su dato, o directamente su dato, o se elimina directamente de memoria.
            #Esto dependera de lo que se desee hacer con el nodo expulsado de la pila
            #En este caso se eliminara de memoria
            del peek
            #Decrementamos el valor de la longitud de la pila
            self.size -= 1
    
    #Funcion para saber cual es el elemento que esta en la cima de la pila
    #Complejidad Temporal: O(1)
    def peek(self):
        #Verificamos que la lista no este vacia
        if self.size > 0 :
            #Aca como se esta trabajando con nodos, se puede retornar el nodo o directamente el dato que contiene
            #En este caso, retornaremos el dato que contiene el nodo
            return self.top.next.data

#Ahora haremos la funcion que halla el NGE de cada elemento del arreglo pero usando dos ciclos anidados, es decir, la solucion con complejidad Cuadratica
#Complejidad Temporal: O(n^2)
#Esta funcion recibira el vector sobre el cual se iterara para hallar los maximos siguientes de cada elemento
def NGEArrays(arr):
    #Lo primero que debemos de hacer es crear una variable la cual ira almacenando el elemento mayor proximo de cada posicion del arreglo. Inicialmente, este valdra cero como para decir que no tiene ningun valor actualmente, ya que este se ira actualizando a lo largo del algoritmo 
    Next = 0
    #Ahora lo que tendremos que hacer es iterar toda la longitud del arreglo
    for i in range(len(arr)):
        #Una vez dentro del arreglo, antes de hacer el ciclo interno, debemos de indicar que ahora la variable Next valdra -1.
        #Esto es con el motivo, que con el ciclo interno se tendra el proposito de encontrar el siguiente mayor del elemento en el que estamos. Sin embargo, si no hay uno mayor, la variable next necesitara tener el valor de -1 como respuesta
        Next = -1
        #Ahora, podemos hacer el ciclo interno. Como bien se sabe , este se encargara de hallar el proximo elemento a la derecha de la posicion que indique el ciclo externo . Por ello, debemos de iniciar este ciclo, desde la posicion siguiente de la que indica el ciclo externo 
        for j in range((i+1), len(arr)):
            #Dentro del ciclo interno , lo que debemos de hacer es que si el elemento que indica la posicion del ciclo externo es menor al elemento que indica la posicion del ciclo interno, quiere decir que ese sera el elemento proximo mayor, y este debera de ser almacenado en la variable Next que creamos
            #Ademas , alli ya podemos parar el ciclo interno ya que de nada sirve seguir iterando sobre el resto del arreglo
            if arr[i] < arr[j]:
                Next = arr[j]
                break
        #Ahora bien, ya que estamos afuera del ciclo  interno , ya tenemos guardado en la variable Next el valor del proximo mayor elemento de la posicion que estamos. Por ello , podemos pasar a imprimir en que elemento estamos y cual es su siguiente
        print(f"Elemento actual : {arr[i]} . Su proximo elemento mayor es : {Next}")

#Ahora veremos un algoritmo optimizado del problema anterior utilizando pilas monotonas
#Complejidad Temporal: O(n)
def NGEStack(arr):
    #Lo primero que debemos de hacer es crear la pila que utilizaremos para el algoritmo 
    stack = Stack()
    #Ahora lo que haremos sera insertar en la pila el primer elemento del vector
    stack.push(arr[0])
    #Ahora deberemos de iterar sobre el arreglo, En este caso partiendo desde el indice 1 ya que el dato del elemento cerp ya esta insertado en la pila
    for i in range(1, len(arr)):
        #Ahora, debemos de verificar que si la pila esta vacia deberemos de insertar el dato en la posicion que nos encontremos actualmente, y continuar en la siguiente iteracion
        #Es importante validar estos primero ya que en proximos pasos del algoritmo se podran realizar varias veces pops de la pila, por lo que no sabremos si en algun momento esta quedara vacia. De ahi, el motivo de que insertemos el dato en la posicion en que nos encontremos si esto ocurre
        if stack.size == 0 :
            stack.push(arr[i])
            continue
        #Luego de lo anterior, significa que la pila no se encuentra vacia, por lo que debemos de comparar el dato en la posicion en la que estamos del vector, con el dato que esta en el Top de la pila. 
        #Si el dato en la cima de la pila es menor al dato en la posicion en la que nos encontramos, significara que el siguiente proximo del dato que esta en la cima de la pila, sera el dato en el que nos encontremos.
        #Por ello, debemos imprimir que el siguiente de la cima de la pila, es el dato en el que nos encontramos y luego hacer un pop de la pila para que los otros datos que se encuentran en ella , pueden seguir comparandose con el dato de la posicion en la que nos encontramos
        while stack.size > 0 and stack.peek() < arr[i]:
            print(f"El siguiente proximo mayor de {stack.peek()} es {arr[i]}")
            stack.pop()
        #Al salir del ciclo anterior, significa que el dato en la cima de la pila es mayor al dato en el que nos encontramos o que la pila esta vacia
        #Por ello, debemos de insertar en la pila ahora el dato en el que nos encontramos
        stack.push(arr[i])
        #Con esto, podemos continuar con las iteraciones del ciclo principal
    #Ahora, afuera del ciclo principal, podran quedar varios datos aun en la pila. Estos datos significan que no tiene proximo mayor , por lo tanto la respuesta de estos sera -1. 
    #Por ellos, debemos de iterar en la pila imprimiendo que el siguiente proximo mayor de la cima de la pila sera -1 , e ir haciendo pops en la pila hasta que esta quede vacia
    while stack.size > 0 :
        print(f"El siguiente proximo mayor de {stack.peek()} es {-1}")
        stack.pop()

#''''''''''''''''''''''''''''''''''''''''''''''''''''''''''

File: Stacks/#4.Monotonic-Stacks/C++/NGE.cpp
/*
En este archivo veremos uno de los problemas as comunes que se pueden resolver de una forma optima utilizando pilas monotonas.
Este problema se trata  que dado un arreglo de elementos numericos, hay que hallar cual es el siguiente elemento mayor en el arreglo, de cada posicion. Si el elemento no tiene uno mayor a la derecha, sera -1 la respuesta de este. 
Veamos un ejemplo de como seria una entrada y salida de esto:

Input: arr[] = [4,5,2,25]
Output:  4 -> 5
		 5 -> 25
		 2 -> 25
		 25 -> -1

Ahora bien , lo que primero se pensaria al hacer este problema es iterar todo el arreglo por cada elemento de este, buscando proximo mayor de la posicion del arreglo en  la que nos encontremos. Sin embargo, en cuanto a eficiencia es un mal algoritmo ya que su complejidad temporal es cuadratica. 
Por ello, es mas conveniente usar otras estructuras como lo es la pila monotona o un mapa, los cuales aumentando la complejidad espacial a lineal, se puede reducir la complejidad temporal a lineal, ya sea con pilas monotonas o con un mapa.

Sin embargo, en este ocasion veremos como resolver este problema de la forma tipica, es decir, con dos ciclos, y con pilas monotonas, para mirar las diferencias entre ambos algoritmos, y asi ver una de las tantas aplicaciones que nos permite esta variante de las pilas. En este caso, implementaremos nuestra pila mediante una lista enlazada. 
*/


#include<bits/stdc++.h>
using namespace std;

//Lo primero que necesitamos hacer es crear la estructura nodo
//Para esto crearemos una clase llamada Node, la cual contendra los atributos que contiene un nodo: El valor y el puntero al siguiente nodo
class Node{
    public:
        int data;
        Node *next;
        Node(){this->data = 0 ; this->next = NULL;}
        Node(int data){this->data = data ; this->next = NULL;}
};

//Ahora crearemos la estructura Lista , en este caso, Stack
//Si bien en la seccion de listas veiamos que esta estructura tenia un nodo cabeza, en este caso para las pilas, lo llamaremos el nodo top el cual se ira acomodando en el ultimo elemento en insertar de la lista
class Stack{
    public:
        Node *top;
        int size;
        Stack(){this->top = new Node() ; this->size = 0;}
};

//Ahora veremos como retirar el elemento de la pila, es decir, la funcion pop
//Como se vio en la teoria, en una pila, solo se puede ir sacando desde el dato mas es la cima\
//Complejidad Temporal: O(1)
void pop(Stack *S , Node *&top){
    //Verificamos que el Stack no este vacio
    if(S->size > 0){
        //Como bien sabemos, el nodo top , siempre estara apuntando al nodo en la cima de la pila, es decir, al nodo top
        //Por ello, debemos de actualizar el valor de top, ya que debe de apuntar al siguiente del que esta apuntando
        //Para empezar, guardaremos en un nodo auxiliar, el nodo en la cima de la pila
        Node *peek = top->next;
        //Ya guardado el nodo de la cima, haremos que ahora el nodo top apunte al siguiente del nodo de la cima
        top->next = peek->next;
        //Desvinculamos el que era el nodo de la cima del stack haciendolo apuntar a nulo
        peek->next = NULL;
        //Ahora bien, con el nodo que estaba en la cima del Stack, hay dos opciones, retornar este para utilizarlo en alguna operacion o eliminarlo directamente de memoria. 
        //Esto dependera de que se quiera hacer con este, ya sea solo expusarlo de la pila y  manipularlo, o directamente eliminarlo
        //En este caso eliminaremos el nodo
        delete peek;
        //Decrementamos el vamor de la pila
        S->size--;
    }   
}

bool isEmpty(Stack *S){
    //Retorna el valor booleano de la siguiente condicion, indicando si es verdadero o falso
    return S->size > 0;
}

//Ahora veremos una funcion la cual retornara cual es el dato que esta en la cima de la pila
//En este caso, como estamos trabajando con nodos, podemos decidir si retornar el nodo en donde se encuentra el dato o directamente el dato que se encuentra en el nodo
//Para este caso, se retornara el dato del nodo
//Complejidad Temporal : O(1)
int peek(Stack *S, Node*&top){
    //Verificamos que la pila no este vacia
    if(S->size > 0){
        //Retornamos el dato del nodo que esta en la cima del Stack
        return top->next->data;
    }
}

//Ahora veremos la funcion de insertar en la pila, es decir, la funcion push
//Complejidad Temporal: O(1)
void push(Stack *S ,Node *&top, int data){
    //Creamos el nuevo nodo a insertar
    Node *new_node = new Node(data);
    //Ahora lo que debemos hacer, es que el nuevo nodo creado, apunte al nodo que apunta top
    new_node->next = top->next;
    //Ahora, ya que hay un nuevo nodo en el Stack, debemos de actualizar el puntero de Top ya que este siempre debe de apuntar al nodo que esta en la cima de la pila, es decir, el ultimo nodo en ser insertado
    top->next = new_node;
    //Incrementamos el valor del tamano del Stack
    S->size++;
}

//La primera funcion para resolver el problema en cuestion, sera iterando sobre todos los elementos del arreglo por cada elemento de este
//Complejidad Temporal: O(n^2)
//Esta funcion recibira el arreglo con los datos, y el tamano del arreglo .Luego,lo que hara al final es imprimir cada elemento del arreglo con su proximo mayor
void NGEArrays(int arr[], int n ){
    //Lo primero que se hara sera crear una varible la cual ira guardando el proximo elemento de cada elemento del vector
    int next;
    //Ahora, iteraremos hasta la longitud del arreglo
    for(int i = 0 ; i < n ; ++i){
        //Ahora bien , inicializaremos a next en -1 asumiendo que este no tendra valor mayor proximo 
        next = -1;
        //Ahora, como debemos de encontrar el elemento mayor a la derecha de la posicion sobre la que estamos, debemos de iterar todo el arreglo, empezando desde la posicion siguiente de la que nos encontramos
        for(int j = i + 1 ; j < n ; ++j){
            //Ahora debemos de verificar que si la posicion en i , es decir, la posicion sobre la que estamos parados, es menor a la posicion en j que es la que esta iterando sobre el resto de los elementos del vector, significa que ese dato en j es el proximo mayor al dato sobre el que estamos iterando. De modo, que la variable next guardara ese valor que esta en j.
            if (arr[i] < arr[j]){
                next = arr[j];
                break;
            }
        }
        //Ahora, que estamos afuera del ciclo interno, ya tenemos guardado el elemento mayor proximo de la posicion que indica el ciclo externo.
        //Por ello, podemos proceder a  imprimir el elemento de la posicion actual con su proximo 
        cout<<"Elemento actual: "<<arr[i]<<" . Su elemento mayor proximo es : "<<next<<endl;
    }
}

//Ahora veremos como optimizar el anterior algoritmo con pilas monotonas
//Complejidad Temporal: O(n)
//Esta funcion recibira el arreglo con los datos y la longitud de este 
void NGEStack(int arr[], int n){
    //Crearemos el Stack en donde se iran insertando los datos
    Stack *s = new Stack();
    //Insertamos en la pila el primer elemento del arreglo
    push(s, s->top , arr[0]);
    //Iteramos sobre todo el vector, empezando desde el indice uno ya que el elemento del indice cero ya lo insertamos en la pila
    for(int i = 1 ; i < n ; ++i){
        //Lo primero que debemos de verificar es si la pila esta vacia. Esto puede ocurrir ya que mas adelante en el codigo se podran hacer varios pops en la pila dejando la posibilidad de vaciarla en algunos casos
        //Si esto ocurre , lo que debemos de hacer es insertar en la pila, el dato de la posicion en la que nos encontramos y pasar a la siguiente iteracion
        if(s->size == 0){
            push(s, s->top , arr[i]);
            continue;
        }
        //Ahora lo que debemos de hacer es que mientras la pila no este vacia y el dato en la posicion que estamos es mayor al elemento Top de la pila, haremos pop en la pila, y el proximo mayor del elemento popeado de la pila sera la el dato de la posicion en la que estamos
        while(s->size > 0  && arr[i] > peek(s, s->top)){
            //Si las condiciones anteriores se cumplen, significa que el proximo del elemento Top de la pila , es el dato de la posicion en la que nos encontranos
            //Por ello, ya podemos imprimir el dato de la cima de la pila, con su siguiente proximo 
            cout<<"El siguiente mayor de "<<peek(s , s->top)<< " es "<< arr[i]<<endl;
            //Luego de imprimir,debemos de hacer pop en la pila, para que se puede comparar el siguiente elemento de la pila con el dato de la posicion en la que estamos
            pop(s, s->top);
        }
        //Luego de lo anterior, una vez que ya el dato en el que estamos sea menor , al dato de la cima de la pila, tendremos que hacerle un push a este para insertarlo en la pila y comparar con la siguiente posicion del arreglo
        push( s, s->top , arr[i]);
    }
    //Una vez terminado en el ciclo anterior, puede que en la pila queden datos. Esto significa que estos datos no tenian un proximo mayor, por lo que debemos de imprimir que su siguiente mayor es -1, y hacerles pop en la pila
    while(s->size > 0){
        //Imprimimos que el siguiente proximo mayor del dato en la cima de la pila es -1
        cout<<"EL siguiente mayor de "<<peek(s, s->top)<<" es "<<-1<<endl;
        //Hacemos pop en la pil
        pop(s, s->top);
    }
}

//------------------------------------------------------------------------------------------